package/src/lsif.ts

```diff
@@ -0,0 +1,229 @@
+    }
+
+    if (lsifDocs.has(doc.uri)) {
+        return lsifDocs.get(doc.uri)!
+    }
+
+    const url = new URL('.api/lsif/exists', sourcegraph.internal.sourcegraphURL)
+    url.searchParams.set('repository', repositoryFromDoc(doc))
+    url.searchParams.set('commit', commitFromDoc(doc))
+    url.searchParams.set('file', pathFromDoc(doc))
+
+    const hasLSIFPromise = (async () => {
+        try {
+            // Prevent leaking the name of a private repository to
+            // Sourcegraph.com by relying on the Sourcegraph extension host's
+            // private repository detection, which will throw an error when
+            // making a GraphQL request.
+            await queryGraphQL({
+                query: `query { currentUser { id } }`,
+                vars: {},
+                sourcegraph,
+            })
+        } catch (e) {
+            return false
+        }
+        const response = await fetch(url.href, {
+            method: 'POST',
+            headers: new Headers({
+                'x-requested-with': 'Sourcegraph LSIF extension',
+            }),
+        })
+        if (!response.ok) {
+            throw new Error(`LSIF /exists returned ${response.statusText}`)
+        }
+        return await response.json()
+    })()
+
+    lsifDocs.set(doc.uri, hasLSIFPromise)
+
+    return hasLSIFPromise
+}
+
+async function hover(
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+): Promise<sourcegraph.Hover | null> {
+    const hover: LSP.Hover | null = await queryLSIF({
+        doc,
+        method: 'hover',
+        path: pathFromDoc(doc),
+        position,
+    })
+    if (!hover) {
+        return null
+    }
+    return convertHover(sourcegraph, hover)
+}
+
+async function definition(
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+): Promise<sourcegraph.Definition | null> {
+    const body: LSP.Location | LSP.Location[] | null = await queryLSIF({
+        doc,
+        method: 'definitions',
+        path: pathFromDoc(doc),
+        position,
+    })
+    if (!body) {
+        return null
+    }
+    const locations = Array.isArray(body) ? body : [body]
+    return convertLocations(
+        sourcegraph,
+        locations.map((definition: LSP.Location) => ({
+            ...definition,
+            uri: setPath(doc, definition.uri),
+        }))
+    )
+}
+
+async function references(
+    doc: sourcegraph.TextDocument,
+    position: sourcegraph.Position
+): Promise<sourcegraph.Location[] | null> {
+    const locations: LSP.Location[] | null = await queryLSIF({
+        doc,
+        method: 'references',
+        path: pathFromDoc(doc),
+        position,
+    })
+    if (!locations) {
+        return []
+    }
+    return convertLocations(
+        sourcegraph,
+        locations.map((reference: LSP.Location) => ({
+            ...reference,
+            uri: setPath(doc, reference.uri),
+        }))
+    )
+}
+
+export type Maybe<T> = { value: T } | undefined
```

Comment:
What is the benefit of this over a simple union type `T | undefined`?

Reply:
`{ value: T }` lets a consumer disambiguate between the two variants, and works for all `T` (including `undefined`).